test/index.js

```diff
@@ -1,223 +1,420 @@
-                setTimeout(checkDead, 400); // 400 = timeout + 100ms spawntime
-            }), 10);
-        });
-        function checkDead() {
-            t.equal(lib.balancer.workers.length, 2, "only 2 workers should be active");
-            t.end(); // this test will never end
-        }
-    })
-});
+        it('should properly hang up if exceeds timeout', function(done) {
+            //200ms request vs 100ms timeout = socket trashed
+            request.get('/200')
+                .end(function(err, res) {
+                    assert(reloaded, 'reloaded should be true');
+                    assert.equal(err.message, 'socket hang up');
+                });
 
-runTest("server with arguments", {file: 'server-with-args.js', args: ['fail']}, function(t) {
-    request({url: 'http://localhost:8000'}, function(err) {
-        t.equal(err, 404, 'should get 404 error');
-        t.end();
+            setImmediate(function() {
+                handler.reload();
+                handler.once('ready', function() {
+                    reloaded = true;
+                });
+            });
+
+            intervalHandle = setInterval(function() {
+                if (Object.keys(cluster.workers).length === 2) {
+                    clearInterval(intervalHandle);
+                    return done();
+                }
+                --i;
+                if (i < 0) {
+                    clearInterval(intervalHandle);
+                    return done(new Error('recluster did not gracefully handle the reload'));
+                }
+            }, tryToDoneEvery);
+        });
     });
-});
 
-/**
- * Termination tests
- * 1) dies ungracefully after 1/2 s (lib/server-die-halfsec.js)
- * 2) disconnects  IPC  after 1/2 s (lib/server-disconnect-halfsec.js)
- * 3) msgs 'disconnect' after 1/2 s (lib/server-msg-disconnect-halfsec.js)
- * Then test if
- * 1) they have been replaced quickly
- * 2) they have been killed after a while for (2) and (3)
- *
- * Recommended settings are:
- * respawn: 0.01 - for quick respawns
- * backoff: 0.01 - for quick respawns
- * workers: 2 - to make sure we're not testing only for one
- * timeout: 0.3 - to be able to test if timeout works
- * before the replacement tries to signal that its dead
- *
- * I know that timing-based tests are not perfect, but I have no better
- * idea at the moment.
- */
+    //@FIXME: help me here, no idea what we are supposed to test
+    //or even what the proper behavior SHOULD be
+    describe('undying', function() {
+        this.timeout(20000);
+
+        it('old undying workers should not respond', function(done) {
+            //copy unclean to temp
+            try {
+                fs.copySync(
+                    path.join(__dirname, 'lib', 'server-unclean.js'),
+                    path.join(__dirname, 'lib', 'server-temp.js')
+                );
+            } catch (err) {
+                done(err);
+            }
```

Comment:
It copies to server-temp, spin up recluster using server-temp. Copy the 'clean' server over server-temp, reload, and ...

I don't know what is the expected behavior, so I tested at least that the previous unclean server is no longer answering, but I expected that because I waited for the reload callback that some workers were ready to answer.

They weren't, at this step I usually get 50 ECONNREFUSED. Shouldn't the new worker(s) be able to send a correct response because they are listening ?